### DISCUSSION TITLE
CLI and FRR vtysh integration

### DISCUSSION THREAD

**Discussion by minexn** [kernelkit/infix#153] [Q&A] [answered]
Are there any plans to integrate the configuration of FRR into the Infix CLI?

At the moment, the configuration is done through the vtysh CLI.

**Comment by troglobit** [comment, marked as answer]
## Short answer

Yes

## Slightly longer answer

FRR comes with a "northbound" integration with Sysrepo. Ideally, we would be able to leverage that for all our dynamic routing needs in Infix. We would then load FRR's YANG models into Sysrepo, configuration could be done via any of our front-ends (CLI or NETCONF), and FRR would apply the changes to the relevant protocol implementations.

However, the models they have are very "clunky" and complex. This is the JSON document needed to add a static interface route to the loopback interface:

```json
{
  "frr-routing:routing": {
    "control-plane-protocols": {
      "control-plane-protocol": [
        {
          "type": "frr-staticd:staticd",
          "name": "staticd",
          "vrf": "default",
          "frr-staticd:staticd": {
            "route-list": [
              {
                "prefix": "10.0.0.0/8",
                "afi-safi": "frr-routing:ipv4-unicast",
                "path-list": [
                  {
                    "table-id": 0,
                    "distance": 1,
                    "frr-nexthops": {
                      "nexthop": [
                        {
                          "nh-type": "ifindex",
                          "vrf": "default",
                          "gateway": "",
                          "interface": "lo"
                        }
                      ]
                    }
                  }
                ]
              }
            ]
          }
        }
      ]
    }
  }
}
```

Finally, **the Sysrepo module no longer seem to be maintained**. https://github.com/FRRouting/frr/issues/12652#issuecomment-1402223353

## Conclusion

Given that...
- ...the existing code base is treated as "experimental" by the FRR developers...
- ...we do not have the time to invest in bringing the code up to par...
- ...the models are not suitable for "direct" consumption by users (i.e. in the CLI)...

...we have concluded that the sysrepo module in FRR is _not_ the way forward for integrating FRR in Infix at this point in time.

Instead we are looking into using standard IETF models, or even our own models, to bring the same functionality to our users.  The first couple of models will be static routes and OSPF (basic functionality first).

**Comment by mattiaswal** [comment]
As of Infix 2023.12, there are support for static routes (IPv4 and IPv6) and basic OSPF configurations @minexn. It is implemented using the IETF models as discussed above, please test it out. Take a look at the deviations to see whats currently is supported of the YANG models (ietf-routing, ietf-ospf) since all not supported parts should been deviated as not supported.
